File: src/main/java/com/hileco/cortex/instructions/jumps/JUMP.java
package com.hileco.cortex.instructions.jumps;

import com.hileco.cortex.context.VirtualMachine;
import com.hileco.cortex.context.ProgramContext;
import com.hileco.cortex.context.ProgramZone;
import com.hileco.cortex.instructions.ProgramException;
import com.hileco.cortex.instructions.StackParameter;
import lombok.EqualsAndHashCode;

import java.math.BigInteger;
import java.util.Collections;
import java.util.List;

import static com.hileco.cortex.context.ProgramZone.INSTRUCTION_POSITION;
import static com.hileco.cortex.instructions.ProgramException.Reason.STACK_TOO_FEW_ELEMENTS;

@EqualsAndHashCode(callSuper = true)
public class JUMP extends JumpingInstruction {

    public static final StackParameter ADDRESS = new StackParameter("address", 0);

    @Override
    public void execute(VirtualMachine process, ProgramContext program) throws ProgramException {
        var stack = program.getStack();
        if (stack.size() < 1) {
            throw new ProgramException(program, STACK_TOO_FEW_ELEMENTS);
        }
        var nextInstructionPosition = new BigInteger(stack.pop()).intValue();
        this.performJump(program, nextInstructionPosition);
    }

    @Override
    public List<Integer> getStackAdds() {
        return List.of();
    }

    @Override
    public List<ProgramZone> getInstructionModifiers() {
        return Collections.singletonList(INSTRUCTION_POSITION);
    }

    @Override
    public List<StackParameter> getStackParameters() {
        return List.of(ADDRESS);
    }

    @Override
    public String toString() {
        return "JUMP";
    }
}

File: src/main/java/com/hileco/cortex/instructions/jumps/JUMP.kt
package com.hileco.cortex.instructions.jumps

import com.hileco.cortex.constraints.expressions.Expression
import com.hileco.cortex.instructions.ProgramException
import com.hileco.cortex.instructions.ProgramException.Reason.STACK_UNDERFLOW
import com.hileco.cortex.instructions.StackParameter
import com.hileco.cortex.vm.ProgramZone
import com.hileco.cortex.vm.ProgramZone.INSTRUCTION_POSITION
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.VirtualMachine
import com.hileco.cortex.vm.symbolic.SymbolicProgramContext
import com.hileco.cortex.vm.symbolic.SymbolicVirtualMachine
import java.math.BigInteger

class JUMP : JumpingInstruction() {
    override val instructionModifiers: List<ProgramZone>
        get() = listOf(INSTRUCTION_POSITION)

    override val stackParameters: List<StackParameter>
        get() = listOf(ADDRESS)

    override fun execute(virtualMachine: VirtualMachine, programContext: ProgramContext) {
        if (programContext.stack.size() < 1) {
            throw ProgramException(STACK_UNDERFLOW)
        }
        val nextInstructionPosition = BigInteger(programContext.stack.pop()).toInt()
        performJump(programContext, nextInstructionPosition)
    }

    override fun execute(virtualMachine: SymbolicVirtualMachine, programContext: SymbolicProgramContext) {
        if (programContext.stack.size() < 1) {
            throw ProgramException(STACK_UNDERFLOW)
        }
        val targetExpression = programContext.stack.pop() as? Expression.Value
                ?: throw UnsupportedOperationException("Jumps to non-concrete targets are not supported for symbolic execution")
        performJump(programContext, targetExpression.constant.toInt())
    }

    companion object {
        val ADDRESS = StackParameter("address", 0)
    }
}
